src/v/cloud_storage_clients/client_pool.h

```diff
@@ -116,11 +129,18 @@ class client_pool : public ss::weakly_referencable<client_pool> {
     http_client_ptr make_client() const;
     void release(http_client_ptr leased);
 
+    /// Return number of clients which wasn't utilized
+    size_t anounce_availability() const;
+    bool borrow_one();
+    void return_one();
+
     ///  Wait for credentials to be acquired. Once credentials are acquired,
     ///  based on the policy, optionally wait for client pool to initialize.
     ss::future<> wait_for_credentials();
 
-    const size_t _max_size;
+    /// Configured capacity per shard
+    const size_t _capacity;
+    moving_max<size_t, 16> _availability_stat;
```

Comment:
I don't understand why we are using historical data here. Could you elaborate?

Reply:
The idea is that if the all connections were borrowed from the shard or the shard was using all its connections other shards won't be able to borrow from it for some time.

Now I'm thinking that maybe this wasn't a good idea because if one shard is underutilised and another is overutilized the second shard will borrow all connections once and then it won't be able to do this for some time.